docs/design.md

```diff
@@ -0,0 +1,508 @@
+metadata:
+  name: remote-example
+spec:
+  forProvider:
+    playbook: sample_namespace.sample_collection.sample_playbook
+  providerConfigRef:
+    name: provider-config-example
+```
+
+If multiple playbooks are listed using `spec.forProvider.playbooks`, they will be run sequencially one after another. At the time of writing this document, only single playbook execution is supported.
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: remote-example
+spec:
+  forProvider:
+    playbooks:
+      - sample_namespace.sample_collection.sample_playbook
+      - sample_namespace.sample_collection.another_sample_playbook
+  providerConfigRef:
+    name: provider-config-example
+```
+
+The collection that includes the playbook is defined as a requirement in `ProviderConfig` and  retrieved from a remote place.
+
+When using inline playbook, besides builtin Ansible modules, you can also include roles or playbooks from collections that are defined as requirements in `ProviderConfig` and retrieved from remote place. For example, below inline playbook is to call `nginx` role in `nginxinc.nginx_core` collection:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: inline-example
+spec:
+  forProvider:
+    playbookInline: |
+      ---
+      - hosts: all
+        collections:
+          - nginxinc.nginx_core
+        tasks:
+          - name: Install NGINX
+            include_role:
+              name: nginx
+  providerConfigRef:
+    name: provider-config-example
+```
+
+The corresponding collection is defined as requirement in `ProviderConfig`:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: ProviderConfig
+metadata:
+  name: example
+spec:
+  requirements: |
+    ---
+    collections:
+      - name: nginxinc.nginx_core
+        version: 0.5.0
+```
+
+## Passing Variables
+
+Ansible uses variables to manage differences among systems on which Ansible operates, so it can run roles or playbooks on multiple systems using single command. Ansible provider allows you to pass those differences into Ansible run through `vars` field when define `AnsibleRun` resource. 
+
+Here is an example to define some variables including a simple variable, a list variable, and a dictionary variable. These variables will be passed into Ansible role `sample_namespace.sample_role` for it to consume:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: remote-example
+spec:
+  forProvider:
+    role: sample_namespace.sample_role
+    vars:
+      foo: value1
+      bar:
+      - value2
+      - value3
+      baz:
+        field 1: value1
+        field 2: value2
+  providerConfigRef:
+    name: provider-config-example
+```
+
+The way to reference these variables in Ansible roles or playbooks is exactly the same as it is defined in Ansible docs using Jinja2 syntax:
+
+```
+{{ foo }}
+{{ bar[0] }}
+{{ baz['field1'] }}
+{{ baz.field2 }} 
+```
```

Comment:
ah Sorry, it is good for me 
I misunderstood the example :)

Reply:
Sorry for the lack of details above, What I am saying is allowing multiple ways to declare var in the AnsibleRun spec is confusing
here we have 3 declaration type : 
```
apiVersion: ansible.crossplane.io/v1alpha1
kind: AnsibleRun
metadata:
  name: remote-example
spec:
  forProvider:
    role: sample_namespace.sample_role
    vars:
      foo: value1
      bar:
      - value2
      - value3
      baz:
        field 1: value1
        field 2: value2
  providerConfigRef:
    name: provider-config-example
```

I think we can just allow a specific declaration type and format `key/value` as required on controller side as we did here https://github.com/multicloudlab/crossplane-provider-ansible/blob/main/pkg/runnerutil/runnerutil.go#L59

```
apiVersion: ansible.crossplane.io/v1alpha1
kind: AnsibleRun
metadata:
  name: remote-example
spec:
  forProvider:
    role: sample_namespace.sample_role
    vars:
      - foo: value1
      - bar: value2
  providerConfigRef:
    name: provider-config-example
```

Reply:
> allowing multiple ways to declare var in the AnsibleRun spec is confusing ...

Got it. Actually, this is how Ansible vars are supported, more details can be found at [here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html). A couple of reasons that I am adding these variations here are that:
- Just follow the same way as Ansible does and align w/ it consistently to help the Ansible provider adoption smoothly for existing Ansible users.
- It is not always sufficient to express vars using key/value pair. Some times, people may ask for vars in the form of list or dictionary, in which case, it is similar to ArgoCD that supports helm values: For those expressed as list or dictionary in helm values.yaml, on Argo CD UI, because of its limitation that only supports key/value form of UI input, it has to be expressed something as below:
```
# list
list[0].foo: value1
list[1].bar: value2

# dictionary
baz.field1: value1
baz.field2: value2
```
This may also be acceptable in our case or at least at our first few releases. We can enhance it iteratively :-)

Reply:
Sounds good :).
helm is at a higher level compared to a kubernetes CR, precisely its role is to bring this  templating system(s) (e.g list and dictionnary vars) to a k8s resource